README

README for Tk::Tree
===================

This is a port of the Tix widget "TixTree" to Perl/Tk.

Tree is now part of the Tk core code.  This is a patch off of Tk800.0_02.
I haven't tested it, but it will probably work with Tk402 as well.

Please note that a Tree is not Scrolled like in Tix.  If you want it
scrolled use $w->Scrolled( qw/Tree .../ ) or $w->ScrlTree( qw/.../ )

This release includes:

    Tree.pm             The Tree class.
    DirTree.pm          The DirTree class.
    perl-tix-tree       Perl version of Tix4.1.0/demos/samples/Tree.tcl

Enjoy,
Chris Dean
20 March 1998



Changes
Revision history for Perl extension Tk::Tree.

4.75  2007-11-17 Slaven Rezic <srezic@cpan.org>
      * DirTree
        * locale-specific sorting
        * convert utf8 characters back to something perl's filesystem
          functions understand
      * removed t/Tree_Manifest.t -- make distcheck should be enough

4.74  2007-11-16 Slaven Rezic <srezic@cpan.org>
      * (hopefully) fixed DirTree problem on Windows, spotted by
        Martin Thurn (http://rt.cpan.org/Ticket/Display.html?id=30781)
      * a couple of more tests for DirTree

4.73  2007-11-08 Slaven Rezic <srezic@cpan.org>
      * synchronized with version found in Tk SVN repository
        * solved problems with Tk::DirTree under Windows
        * minor test fixes
      * internal change: Makefile.PL has a rule for detecting diffs
        between Tk-Tree and Tk

4.72  2007-11-04 Slaven Rezic <srezic@cpan.org>
      * restored 5.005 compatibility
      * reordered this file

4.71  2007-11-03 Slaven Rezic <srezic@cpan.org>
      * synchronized with version found in Tk SVN repository:
        * Pod found in Tree.pm split into own file
        * restored original indentation (NI-S style) and
          line endings (Unix) 
        * fixed Tk::Widget bug also in Tk::DirTree
        * removed Pod::Coverage test, because it does not fit into
          classic Tk modules (e.g. Populate never needs to be
          documented in a Tk system)
        * Tree.pod now follows the Tk version, which seems to
          be more modern
        * moved examples directory one up, to prevent from installing
          into the site directory

4.7  Wed Feb 07 15:47:24 2007 Renee Baecker <module@renee-baecker.de>
      * fixed the Tk::Widget-Bug, thanks to Slaven Rezic

4.6  Mon Dec 18 10:47:24 2006 Renee Baecker <module@renee-baecker.de>
      * fixed some bugs in Tests

4.5  Mon Oct 02 11:07:14 2006 Renee Baecker <module@renee-baecker.de>
      * added pseudo documentation for some methods
      * added tests: Pod, Pod_Coverage, Manifest

4.4
      * Minor bugfixes

4.3
      * Minor bugfixes

4.2  Tue Nov 15 05:24:52 2005 Renee Baecker <module@renee-baecker.de>
      * Tk/Tree.pm: child_entries implemented
      
4.1  Fri Nov 04 10:59:23 2005 Renee Baecker <module@renee-baecker.de>
      * Tk/Tree.pm: add_pathimage added and code of module changed
      
3.00401  Fri Mar 20 11:45:54 1998  Chris Dean  <ctdean@cogit.com>

      * Tk/DirTree.pm: Created.

      * HList.pm.patched: Starting from HList.pm in Tk800.0_02, removed
        references to EventType.  The -indicatorcmd callback takes the
        EventType as an additional argument.

      * Tk/Tree.pm: Starting from Tree.pm in Tk800.0_02, expect
        -indicatorcmd to have an additional arg.

      * examples/perl-tix-dirtree: Created.

0.05  Mon Jan 12 13:21:41 1998  Chris Dean  <ctdean@cogit.com>

      * Tk/Tree.pm: Version 0.05

      * Tk/VTree.pm: From Achim -- A VTree isa HList, it doesn't have
        a component that is a Scrolled HList.

      * Tk/Tree.pm: From Achim -- Since a Tree is not scrolled by
        default add the Tk::Widget::ScrlTree method.

      * README: Trees are not Scrolled by default.

      * examples/adjuster: Another example

      * examples/perl-tix-*tree: Use ScrlTree

      * Tk/Tree.pm: Mention that Trees are not Scrolled in the docs.

0.04  Mon Dec  8 13:00:46 1997  Chris Dean  <ctdean@cogit.com>

      * Tk/Tree.pm: Version 0.04

      * Tk/VTree.pm: Added VERSION to VTree.pm to help out CPAN.pm.
        Otherwise CPAN.pm's `r' command gets confused.  (From Achim
        Bohnet <ach@mpe.mpg.de>)

0.03  Mon Dec  8 09:26:32 1997  Chris Dean  <ctdean@cogit.com>

      * Tk/Tree.pm: Version 0.03

      * Tk/Tree.pm: More doc changes from Achim.  Removed some fo the
        subwidget section.

      * Tk/VTree.pm: Verify correct HList.pm version.

      * test.pl: Check for patched HList.pm.

      * Makefile.PL: Use new HList.pm if needed.

      Fri Dec  5 10:46:19 1997  Chris Dean  <ctdean@cogit.com>

      * MANIFEST: Removed folder.xpm (already in Tk402.003), and Tree.n
        (the pod info supersede it).

      * Tk/Tree.pm: Moved Tree.pod to end of Tree.pm

      * Tk/Tree.pod: Removed some Tix leftovers via patch
        from Achim Bohnet <ach@mpe.mpg.de>.


0.02  Thu Dec  4 17:51:01 PST 1997
        - Fixed package name in Makefile.PL.

0.01  Thu Dec  4 14:32:54 1997
	- original version; created by h2xs 1.18



Makefile.PL
use ExtUtils::MakeMaker;
# See lib/ExtUtils/MakeMaker.pm for details of how to influence
# the contents of the Makefile that is written.
WriteMakefile(
    NAME              => 'Tk::Tree',
    VERSION           => '4.75',
    PREREQ_PM         => {
                           Tk        => 800.024,
                           Tk::HList => 3.040,
                         }, # e.g., Module::Name => 1.1
    ($] >= 5.005 ?     ## Add these new keywords supported since 5.005
      (ABSTRACT_FROM  => 'lib/Tk/Tree.pod', # retrieve abstract from module
       AUTHOR         => 'Chris Dean, Renee Baecker, Slaven Rezic') : ()),
);

sub MY::postamble {
    my $postamble = <<'EOF';

#TKWC=$(HOME)/work/Tk/trunk
TKWC=$(HOME)/work/svk-checkouts/Tk-amd64
MYDIFF=diff -up

diff-tk:
	@$(MYDIFF) t/dirtree.t         $(TKWC)/t/
	@$(MYDIFF) lib/Tk/Tree.pm      $(TKWC)/Tixish/
	@$(MYDIFF) lib/Tk/Tree.pod     $(TKWC)/pod/
	@$(MYDIFF) lib/Tk/DirTree.pm   $(TKWC)/Tixish/
	@$(MYDIFF) lib/Tk/DirTree.pod  $(TKWC)/pod/

pre-dist:
	cpan_check_versions

EOF

    if (defined $ENV{USER} && $ENV{USER} eq 'eserte' && $^O =~ /bsd/i && -f "../../perl.release.mk") {
	$postamble .= <<'EOF';

.include "../../perl.release.mk"
.include "../../perl.cvs.mk"

EOF
    }

    $postamble;
}


lib/Tk/DirTree.pm
package Tk::DirTree;
# DirTree -- TixDirTree widget
#
# Derived from DirTree.tcl in Tix 4.1
#
# Chris Dean <ctdean@cogit.com>

use strict;
use vars qw($VERSION);
$VERSION = sprintf '4.%03d', q$Revision: #14 $ =~ /\D(\d+)\s*$/;

use Tk;
use Tk::Derived;
use Tk::Tree;
use Cwd;
use DirHandle;

use base  qw(Tk::Widget Tk::Derived Tk::Tree);
use strict;

Construct Tk::Widget 'DirTree';


sub Populate {
    my( $cw, $args ) = @_;

    $cw->SUPER::Populate( $args );

    $cw->ConfigSpecs(
        -dircmd         => [qw/CALLBACK dirCmd DirCmd DirCmd/],
        -showhidden     => [qw/PASSIVE showHidden ShowHidden 0/],
        -image          => [qw/PASSIVE image Image folder/],
        -directory      => [qw/SETMETHOD directory Directory ./],
        -value          => '-directory' );

    $cw->configure( -separator => '/', -itemtype => 'imagetext' );
}

sub DirCmd {
    my( $w, $dir, $showhidden ) = @_;
    $dir .= "/" if $dir =~ /^[a-z]:$/i and $^O eq 'MSWin32';
    my $h = DirHandle->new( $dir ) or return();
    my @names = grep( $_ ne '.' && $_ ne '..', $h->read );
    @names = grep( ! /^[.]/, @names ) unless $showhidden;
    return( @names );
}

*dircmd = \&DirCmd;

sub fullpath
{
 my ($path) = @_;
 my $cwd = getcwd();
 if (CORE::chdir($path))
  {
   $path = getcwd();
   CORE::chdir($cwd) || die "Cannot cd back to $cwd:$!";
  }
 else
  {
   warn "Cannot cd to $path:$!"
  }
 return $path;
}

sub directory
{
    my ($w,$key,$val) = @_;
    # We need a value for -image, so its being undefined
    # is probably caused by order of handling config defaults
    # so defer it.
    $w->afterIdle([$w, 'set_dir' => $val]);
}

sub set_dir {
    my( $w, $val ) = @_;
    my $fulldir = fullpath( $val );

    my $parent = '/';
    if ($^O eq 'MSWin32')
     {
      if ($fulldir =~ s/^([a-z]:)//i)
       {
        $parent = $1;
       }
     }
    $w->add_to_tree( $parent, $parent)  unless $w->infoExists($parent);

    my @dirs = ($parent);
    foreach my $name (split( /[\/\\]/, $fulldir )) {
        next unless length $name;
        push @dirs, $name;
        my $dir = join( '/', @dirs );
	$dir =~ s|^//|/|;
        $w->add_to_tree( $dir, $name, $parent )
            unless $w->infoExists( $dir );
        $parent = $dir;
    }

    $w->OpenCmd( $parent );
    $w->setmode( $parent, 'close' );
}
*chdir = \&set_dir;


sub OpenCmd {
    my( $w, $dir ) = @_;

    my $parent = $dir;
    $dir = '' if $dir eq '/';
    foreach my $name ($w->dirnames( $parent )) {
        next if ($name eq '.' || $name eq '..');
        my $subdir = "$dir/$name";
        next unless -d $subdir;
        if( $w->infoExists( $subdir ) ) {
            $w->show( -entry => $subdir );
        } else {
            $w->add_to_tree( $subdir, $name, $parent );
        }
    }
}

*opencmd = \&OpenCmd;

sub add_to_tree {
    my( $w, $dir, $name, $parent ) = @_;

    my $image = $w->cget('-image');
    if ( !UNIVERSAL::isa($image, 'Tk::Image') ) {
	$image = $w->Getimage( $image );
    }
    my $mode = 'none';
    $mode = 'open' if $w->has_subdir( $dir );

    my @args = (-image => $image, -text => $name);
    if( $parent ) {             # Add in alphabetical order.
        foreach my $sib ($w->infoChildren( $parent )) {
            if( $sib gt $dir ) {
                push @args, (-before => $sib);
                last;
            }
        }
    }

    $w->add( $dir, @args );
    $w->setmode( $dir, $mode );
}

sub has_subdir {
    my( $w, $dir ) = @_;
    foreach my $name ($w->dirnames( $dir )) {
        next if ($name eq '.' || $name eq '..');
        next if ($name =~ /^\.+$/);
        return( 1 ) if -d "$dir/$name";
    }
    return( 0 );
}

sub dirnames {
    my( $w, $dir ) = @_;
    my @names = $w->Callback( '-dircmd', $dir, $w->cget( '-showhidden' ) );
    return( @names );
}

{
    package Tk::DirTreeDialog;
    use base qw(Tk::Toplevel);
    Construct Tk::Widget 'DirTreeDialog';

    sub Populate {
	my($w, $args) = @_;
	$w->{curr_dir} = $args->{-initialdir};
	if (!defined $w->{curr_dir}) {
	    require Cwd;
	    $w->{curr_dir} = Cwd::cwd();
	}
	if (defined $args->{-mustexist}) {
	    die "-mustexist is not yet implemented";
	}
	my $title = $args->{-title} || "Choose directory:";
	delete $args->{-popover};

	$w->title($title);
	$w->{ok} = 0; # flag: "1" means OK, "-1" means cancelled

	# Create Frame widget before the DirTree widget, so it's always visible
	# if the window gets resized.
	my $f = $w->Frame->pack(-fill => "x", -side => "bottom");

	my $d;
	$d = $f->Scrolled('DirTree',
			  -scrollbars => 'osoe',
			  -width => 35,
			  -height => 20,
			  -selectmode => 'browse',
			  -exportselection => 1,
			  -browsecmd => sub {
			      $w->{curr_dir} = shift;
			      if ($^O ne 'MSWin32') {
				  $w->{curr_dir} =~ s|^//|/|; # bugfix
			      }
			  },

			  # With this version of -command a double-click will
			  # select the directory
			  -command   => sub { $w->{ok} = 1 },

			  # With this version of -command a double-click will
			  # open a directory. Selection is only possible with
			  # the Ok button.
			  #-command   => sub { $d->opencmd($_[0]) },
			 )->pack(-fill => "both", -expand => 1);
	# Set the initial directory
	exists &Tk::DirTree::chdir ? $d->chdir($w->{curr_dir}) : $d->set_dir($w->{curr_dir});

	$f->Button(-text => 'Ok',
		   -command => sub { $w->{ok} =  1 })->pack(-side => 'left');
	$f->Button(-text => 'Cancel',
		   -command => sub { $w->{ok} = -1 })->pack(-side => 'left');
	$w->OnDestroy(sub { $w->{ok} = -1 });
    }

    sub Show {
	my $w = shift;
	my $old_focus = $w->focusSave;
	my $old_grab = $w->grabSave;
	Tk::catch {
	    $w->grab;
	};
	$w->waitVariable(\$w->{ok});
	my $ret = $w->{ok} == 1 ? $w->{curr_dir} : undef;
	$w->grabRelease if Tk::Exists($w);
	&$old_focus;
	&$old_grab;
	$w->destroy if Tk::Exists($w);
	$ret;
    }
}

1;

__END__

#  Copyright (c) 1996, Expert Interface Technologies
#  See the file "license.terms" for information on usage and redistribution
#  of this file, and for a DISCLAIMER OF ALL WARRANTIES.
#
#  The file man.macros and some of the macros used by this file are
#  copyrighted: (c) 1990 The Regents of the University of California.
#               (c) 1994-1995 Sun Microsystems, Inc.
#  The license terms of the Tcl/Tk distrobution are in the file
#  license.tcl.



lib/Tk/Tree.pm
package Tk::Tree;
# Tree -- TixTree widget
#
# Derived from Tree.tcl in Tix 4.1
#
# Chris Dean <ctdean@cogit.com>
# Changes: Renee Baecker <module@renee-baecker.de>

use vars qw($VERSION);
$VERSION = '4.71'; # $Id: Tree.pm,v 1.2 2007/11/03 21:13:11 eserte Exp $

use Tk ();
use Tk::Derived;
use Tk::HList;
use base  qw(Tk::Derived Tk::HList);
use strict;
use warnings;

Construct Tk::Widget 'Tree';

sub Tk::Widget::ScrlTree { shift->Scrolled('Tree' => @_) }

sub Populate
{
 my( $w, $args ) = @_;

 $w->SUPER::Populate( $args );

 $w->ConfigSpecs(
        -ignoreinvoke => ['PASSIVE',  'ignoreInvoke', 'IgnoreInvoke', 0],
        -opencmd      => ['CALLBACK', 'openCmd',      'OpenCmd', 'OpenCmd' ],
        -indicatorcmd => ['CALLBACK', 'indicatorCmd', 'IndicatorCmd', 'IndicatorCmd'],
        -closecmd     => ['CALLBACK', 'closeCmd',     'CloseCmd', 'CloseCmd'],
        -indicator    => ['SELF', 'indicator', 'Indicator', 1],
        -indent       => ['SELF', 'indent', 'Indent', 20],
        -width        => ['SELF', 'width', 'Width', 20],
        -itemtype     => ['SELF', 'itemtype', 'Itemtype', 'imagetext'],
	-foreground   => ['SELF'],
       );
}

sub autosetmode
{
 my( $w ) = @_;
 $w->setmode();
}

sub add_pathimage
{
 my ($w,$path,$imgopen,$imgclose) = @_;
 $imgopen  ||= "minusarm";
 $imgclose ||= "plusarm";
  
 my $separator = $w->cget(-separator);
  
 $path =~ s/([\.?()|])/\\$1/g;
 $path =~ s/\$/\\\$/g;
 $path =~ s/\\\$$/\$/;
 $path =~ s/\*/[^$separator]+/g;
  
 push(@{$w->{Images}},[$path,$imgopen,$imgclose]);
}

sub child_entries
{
 my ($w,$path,$depth) = @_;
  
 my $level =  1;
 $depth  ||=  1;
 $path   ||= '';
  
 my @children = $w->_get_childinfos($depth,$level,$path);
  
 return wantarray ? @children : scalar(@children);
}

sub _get_childinfos
{
 my ($w,$maxdepth,$level,$path) = @_;
 my @children = $w->infoChildren($path);
 my @tmp;
  
 if($level < $maxdepth)
  {
   for my $child(@children)
    {
     push(@tmp,$w->_get_childinfos($maxdepth,$level +1,$child));
    }
  }
  
 push(@children,@tmp);
  
 return @children;
}

sub IndicatorCmd
{
 my( $w, $ent, $event ) = @_;

 my $mode = $w->getmode( $ent );

 if ( $event eq '<Arm>' )
  {
   if ($mode eq 'open' )
    {
     #$w->_indicator_image( $ent, 'plusarm' );
     $w->_open($ent);
    }
   else
    {
     #$w->_indicator_image( $ent, 'minusarm' );
     $w->_close($ent);
    }
  }
 elsif ( $event eq '<Disarm>' )
  {
   if ($mode eq 'open' )
    {
     #$w->_indicator_image( $ent, 'plus' );
     $w->_open($ent);
    }
   else
    {
     #$w->_indicator_image( $ent, 'minus' );
     $w->_close($ent);
    }
  }
 elsif( $event eq '<Activate>' )
  {
   $w->Activate( $ent, $mode );
   $w->Callback( -browsecmd => $ent );
  }
}

sub close
{
 my( $w, $ent ) = @_;
 my $mode = $w->getmode( $ent );
 $w->Activate( $ent, $mode ) if( $mode eq 'close' );
}

sub open
{
 my( $w, $ent ) = @_;
 my $mode = $w->getmode( $ent );
 $w->Activate( $ent, $mode ) if( $mode eq 'open' );
}

sub getmode
{
 my( $w, $ent ) = @_;

 return( 'none' ) unless $w->indicatorExists( $ent );

 my $img = $w->_indicator_image( $ent );
 if ($img eq "plus" || $img eq "plusarm" || grep{$img eq $_->[2]}@{$w->{Images}})
  {
   return( 'open' );
  }
 return( 'close' );
}

sub setmode
{
 my ($w,$ent,$mode) = @_;
 unless (defined $mode)
  {
   $mode = 'none';
   my @args;
   push(@args,$ent) if defined $ent;
   my @children = $w->infoChildren( @args );
   if ( @children )
    {
     $mode = 'close';
     foreach my $c (@children)
      {
       $mode = 'open' if $w->infoHidden( $c );
       $w->setmode( $c );
      }
    }
  }

 if (defined $ent)
  {
   if ( $mode eq 'open' )
    {
     #$w->_indicator_image( $ent, 'plus' );
     $w->_open($ent);
    }
   elsif ( $mode eq 'close' )
    {
     #$w->_indicator_image( $ent, 'minus' );
     $w->_close($ent);
    }
   elsif( $mode eq 'none' )
    {
     $w->_indicator_image( $ent, undef );
    }
  }
}

sub _open
{
 my ($w,$ent) = @_;
 $w->_indicator_image( $ent, "plus" );
 for my $entry (@{$w->{Images}})
  {
   if($ent =~ $entry->[0])
    {
     $w->_indicator_image( $ent, $entry->[2] );
    }
  }
}

sub _close
{
 my ($w,$ent) = @_;
 $w->_indicator_image( $ent, "minus" );
 for my $entry (@{$w->{Images}})
  {
   if($ent =~ $entry->[0])
    {
     $w->_indicator_image( $ent, $entry->[1] );
    }
  }
}

sub Activate
{
 my( $w, $ent, $mode ) = @_;
 if ( $mode eq 'open' )
  {
   $w->Callback( -opencmd => $ent );
   #$w->_indicator_image( $ent, 'minus' );
   $w->_close($ent);
  }
 elsif ( $mode eq 'close' )
  {
   $w->Callback( -closecmd => $ent );
   #$w->_indicator_image( $ent, 'plus' );
   $w->_open($ent);
  }
 else
  {

  }
}

sub OpenCmd
{
 my( $w, $ent ) = @_;
 # The default action
 foreach my $kid ($w->infoChildren( $ent ))
  {
   $w->show( -entry => $kid );
  }
}

sub CloseCmd
{
 my( $w, $ent ) = @_;

 # The default action
 foreach my $kid ($w->infoChildren( $ent ))
  {
   $w->hide( -entry => $kid );
  }
}

sub Command
{
 my( $w, $ent ) = @_;

 return if $w->{Configure}{-ignoreInvoke};

 $w->Activate( $ent, $w->getmode( $ent ) ) if $w->indicatorExists( $ent );
}

sub _indicator_image
{
 my( $w, $ent, $image ) = @_;
 my $data = $w->privateData();
 if (@_ > 2)
  {
   if (defined $image)
    {
     $w->indicatorCreate( $ent, -itemtype => 'image' )
         unless $w->indicatorExists($ent);
     $data->{$ent} = $image;
     $w->indicatorConfigure( $ent, -image => $w->Getimage( $image ) );
    }
   else
    {
     $w->indicatorDelete( $ent ) if $w->indicatorExists( $ent );
     delete $data->{$ent};
    }
  }
 return $data->{$ent};
}

1;

__END__

#  Copyright (c) 1996, Expert Interface Technologies
#  See the file "license.terms" for information on usage and redistribution
#  of this file, and for a DISCLAIMER OF ALL WARRANTIES.
#
#  The file man.macros and some of the macros used by this file are
#  copyrighted: (c) 1990 The Regents of the University of California.
#               (c) 1994-1995 Sun Microsystems, Inc.
#  The license terms of the Tcl/Tk distrobution are in the file
#  license.tcl.

